src/jdk.jdwp.agent/share/native/libjdwp/classTrack.c

```diff
@@ -81,15 +81,35 @@ cbTrackingObjectFree(jvmtiEnv* jvmti_env, jlong tag)
 struct bag *
 classTrack_processUnloads(JNIEnv *env)
 {
-    debugMonitorEnter(classTrackLock);
-    if (deletedSignatures == NULL) {
-        // Class tracking not initialized, nobody's interested.
-        debugMonitorExit(classTrackLock);
-        return NULL;
-    }
-    struct bag* deleted = deletedSignatures;
-    deletedSignatures = bagCreateBag(sizeof(char*), 10);
-    debugMonitorExit(classTrackLock);
+    struct bag* new_bag = NULL;
+    struct bag* deleted = NULL;
+    jboolean retry = JNI_FALSE;
+    do {
+      // Avoid unnecessary allocations when class track has yet been activated.
+      if (deletedSignatures != NULL) {
+        /* Allocate new bag outside classTrackLock lock to avoid deadlock.
+         *
+         * Note: jvmtiAllocate/jvmtiDeallocate() may be blocked by ongoing safepoints.
+         * It is dangerous to call them (via bagCreateBag/bagDestroyBag())while holding monitor(s),
+         * because jvmti may post events, e.g. JVMTI_EVENT_OBJECT_FREE at safepoints and event processing
+         * code may acquire the same monitor(s), e.g. classTrackLock in cbTrackingObjectFree(),
+         * which can lead to deadlock.
+         */
+        new_bag = bagCreateBag(sizeof(char*), 10);
+      }
+      debugMonitorEnter(classTrackLock);
+      deleted = deletedSignatures;
+      if (deletedSignatures != NULL) {
+        if (new_bag != NULL) {
+          deletedSignatures = new_bag;
+          new_bag = NULL;
+        } else {
+          retry = JNI_TRUE;
+        }
+      }
+      debugMonitorExit(classTrackLock);
+    } while (retry == JNI_TRUE);
+    bagDestroyBag(new_bag);
```

Comment:
One other comment about `classTrack_processUnloads()` in general that also applies to the original version. It seems pretty inefficient. Every time the debug agent gets an event it is called. On almost ever call it is returning and empty back and allocating a new one. It seems a check for `bagSize(deletedSignatures) == 0` and returning NULL if true would help performance. I also believe this can be done outside of the lock (would like David's opinion on this).

Reply:
I am not sure if it is possible, but checking `bagSize(deletedSignatures) == 0` seems to race against `classTrack_reset()` where it does not take `handlerLock` lock.

Reply:
> I am not sure if it is possible, but checking bagSize(deletedSignatures) == 0 seems to race against classTrack_reset() where it does not take handlerLock lock.

I had thought of that too, but I think the way `classTrack_reset()` is called, it is likely not possible for there to be a `classTrack_processUnloads()` also coming in because everything is shut down:

``` 
    threadControl_onDisconnect();
    standardHandlers_onDisconnect();

    /*
     * Cut off the transport immediately. This has the effect of
     * cutting off any events that the eventHelper thread might
     * be trying to send.
     */
    transport_close();
    debugMonitorDestroy(cmdQueueLock);

    /* Reset for a new connection to this VM if it's still alive */
    if ( ! gdata->vmDead ) {
        debugInit_reset(getEnv());    <--- calls classTrack_reset()
    }
```

Reply:
But shutting down debug loop does not seem to have effect on ongoing jvmti callback, e.g. thread 4 in the bug report.

Reply:
The thread 4 callback is triggered due to the debugger having requested `CLASS_UNLOAD` events. This is shutdown by `eventHandler_reset()`, which is called by `debugInit_reset()` before it calls `classTrack_reset()`. So that means by the time we get to `classTrack_reset()`, the thread 4 callback is no longer possible.